src/syntax-highlight/source-handler.js

```diff
@@ -28,6 +28,29 @@ export function getClassBasedOnExtension(element) {
     return '';
 }
 
+/**
+ * Retrieves a class according to the element data-filename or data-path attribute.
+ * This is for specially named universal files like Vagrantfile, Jenkinsfile etc which
+ * have no extension.
+ * Example: "/path/to/file/VagrantFile" would return "language-ruby".
+ * @param  {Element} element An HTML element. Pass anything different and bear the consequences :)
+ * @return {String} The class extracted from the element's file path.
+ */
+export function getClassBasedOnFilename(element) {
+    const filePath = getFilepathFromElement(element);
+    const fileName = getFilename(filePath).toLowerCase();
+
+    if (fileName === 'vagrantfile') {
+        return 'language-ruby';
+    }
+
+    if (fileName === 'jenkinsfile') {
+        return 'language-groovy';
+    }
+
+    return '';
+}
+
```

Comment:
I think it would be nice if we could add support for other universally-named files by just changing the export in `language-ext.js`, instead of having to directly modify the logic of the code. This is actually quite easy to achieve, my suggestion would be to add the following lines to the `getClassBasedOnExtension` function:

```diff
-export function getClassBasedOnExtension(element) {
+export function getLanguageClass(element) {
    const filePath = getFilepathFromElement(element);
    const fileExtension = getExtension(filePath).toLowerCase();

    if (fileExtension in languagesExtensions) {
        return languagesExtensions[fileExtension];
    }

+    const filename = getFilename(filePath).toLowerCase();
+    if (filename in languagesExtensions) {
+        return languagesExtensions[filename];
+    }

    const fileExtensionWithoutDot = fileExtension.slice(1);
    if (fileExtensionWithoutDot in Prism.languages) {
        return `language-${fileExtensionWithoutDot}`;
    }

    return '';
}
```
(Since it now retrieves the language class based not only on the file extension, it would make sense to rename it to `getLanguageClass`.  After that we can delete the new `getClassBasedOnFilename` since we are already taking care of that here.)

This expects the exported object from `language-ext` to have keys for those filenames, so for our specific case it would look like:

```diff
// language-ext.js
module.exports = {
     // ...

+    // Universally-named files
+    vagrantfile: 'language-ruby',
+    jenkinsfile: 'language-groovy'
};
```

And that's actually it, we can then revert the changes in `syntax-highlight.js` and it would work! :D 

What do you think?

Reply:
Also, it would be great if you could add a new test in `source-handler.spec.js` that handles the case of universally-named files!